pkg/controller/mattermostrestoredb/mattermostrestoredb_controller.go

```diff
@@ -0,0 +1,234 @@
+		return reconcile.Result{}, errrors.Wrap(err, "unable to get the statefulset")
+	}
+	if statefulset.Status.ReadyReplicas != 0 {
+		return reconcile.Result{RequeueAfter: time.Second * 3}, fmt.Errorf("Waiting for MySQL Statefulset scale to 0")
+	}
+
+	sel := mattermostv1alpha1.MySQLLabels()
+	opts := &client.ListOptions{LabelSelector: labels.SelectorFromSet(sel)}
+
+	pods := &corev1.PodList{}
+
+	err = r.client.List(context.TODO(), opts, pods)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcile.Result{}, errrors.Wrap(err, "unable to get pod list")
+	}
+	if len(pods.Items) != 0 {
+		return reconcile.Result{RequeueAfter: time.Second * 3}, fmt.Errorf("Waiting for MySQL Statefulset pods scale to 0")
+	}
+
+	reqLogger.Info("MySQL Statefulset are scaled down. Will continue the restore process", "ReadyReplicas", statefulset.Status.ReadyReplicas, "CurrentReplicas", statefulset.Status.CurrentReplicas, "OriginalReplicas", restoreMM.Status.OriginalDBReplicas)
+
+	// Removing all the PVC for MySQL to be able to apply the restore
+	for i := 0; i < int(restoreMM.Status.OriginalDBReplicas); i++ {
+		persistentVolumeClaim := &corev1.PersistentVolumeClaim{}
+		dbPersistentVolClaim := fmt.Sprintf("data-db-mysql-%d", i)
+		reqLogger.Info("Deleting PVC...", "PersistentVolumeClaimName", dbPersistentVolClaim)
+		errGet := r.client.Get(context.TODO(), types.NamespacedName{Name: dbPersistentVolClaim, Namespace: restoreMM.Namespace}, persistentVolumeClaim)
+		if errGet != nil && errors.IsNotFound(errGet) {
+			reqLogger.Info("PVC not found maybe already deleted, skipping", "PersistentVolumeClaimName", dbPersistentVolClaim)
+			continue
+		}
+		if errGet != nil {
+			return reconcile.Result{}, errGet
+		}
+
+		errDelete := r.client.Delete(context.TODO(), persistentVolumeClaim, client.GracePeriodSeconds(0))
+		if errDelete != nil {
+			reqLogger.Error(errDelete, "error deleting the DB PVC", "ClusterInstallation.Namespace", clusterInstallation.Namespace, "ClusterInstallation.Name", clusterInstallation.Name)
+			return reconcile.Result{}, errDelete
+		}
+		reqLogger.Info("PVC deleted", "PersistentVolumeClaimName", dbPersistentVolClaim)
+	}
+
+	// Scale up again to apply the restore
+	clusterInstallation.Spec.Database.Replicas = restoreMM.Status.OriginalDBReplicas
+	if restoreMM.Status.OriginalDBReplicas == 0 {
+		// at least set to one replica
+		clusterInstallation.Spec.Database.Replicas = 1
+	}
+	err = r.client.Update(context.TODO(), clusterInstallation)
+	if err != nil {
+		reqLogger.Error(err, "failed to update the clusterinstallation spec")
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Restore complete", "restoreMM.Namespace", restoreMM.Namespace, "restoreMM.Name", restoreMM.Name)
+	r.setFinished()
+	status := restoreMM.Status
+	status.State = r.state
+	err = r.updateStatus(restoreMM, status, reqLogger)
+	if err != nil {
+		reqLogger.Error(err, "failed to update the clusterinstallation status")
```

Comment:
If we fail to set the status, it looks like we won't retry? Are we okay not returning an error here?

Reply:
Yeah, just like the main clusterinstallation controller, we should return the error if the updateStatus fails. That way, we will reconcile again and will try to update the status until it works.

If we don't make sure the status is updated then we could get into a state where everything is finished, but we don't have status reflecting it.

Reply:
lets do it in a followup